### PR TITLE
Unit test with a set of View Change msg containing Prepared Certs

### PR DESCRIPTION
**Pull request by HristoStaykov**
To fully unit test View Change I think I should put Prepared Certificates in the View Change messages.

### CODE REVIEW

File: bftengine/tests/testViewChange/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      PRIVATE
+      ${bftengine_SOURCE_DIR}/src/bftengine
+      ${concord_bft_tools_SOURCE_DIR})
+
+target_link_libraries(ViewChange_tests PUBLIC
+    GTest::Main
+    util
+    corebft
+    threshsign
+    )
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/CMakeLists.txt]: I've merged your infrastructure PR into master. You can rebase.

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  for (int i = 0; i < N; i++) {
+    delete viewChangeMsgs[i];
+    viewChangeMsgs[i] = nullptr;
+  }
+  delete[] viewChangeMsgs;
+}
+
+}  // end namespace
+
+int main(int argc, char** argv) {
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: Can you also add one test for `calcLBStableForView()` where it returns a proper sequence number, instead of crashing?

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 3]
```diff
@@ -51,6 +135,99 @@ void setUpConfiguration_4() {
...
+
+  SeqNumInfo::init(seqNumInfo_, static_cast<void*>(&replica));
+
+  uint64_t expectedLastValue = 12345;
+  const uint32_t kRequestLength = 2;
+  const uint64_t requestBuffer[kRequestLength] = {(uint64_t)200, expectedLastValue};
+  ViewNum curView = 0;
+  bftEngine::impl::SeqNum AssignedSeqNum = 1;
+
+  auto* CO = new ClientRequestMsg((uint16_t)1,
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: CO -> clientRequest

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 4]
```diff
@@ -99,6 +276,30 @@ TEST(testViewchangeSafetyLogic_test, different_new_views_in_VC_msgs) {
...
 }
 
+TEST(testViewchangeSafetyLogic_test, empty_correct_VC_msgs) {
+  ViewChangeMsg** viewChangeMsgs = new ViewChangeMsg*[N];
+
+  for (int i = 0; i < N; i++) {
+    if (i == 2) {
+      viewChangeMsgs[i] = nullptr;
+    } else {
+      viewChangeMsgs[i] = new ViewChangeMsg(i, 1, 0);
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: same here, can we test with `lastStableSeq > 0`

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 5]
```diff
@@ -51,6 +135,99 @@ void setUpConfiguration_4() {
...
+  DummyReplica replica(replicasInfo, replicaConfig[0]);
+  replica.getInternalThreadPool().start(1);
+
+  SeqNumInfo::init(seqNumInfo_, static_cast<void*>(&replica));
+
+  uint64_t expectedLastValue = 12345;
+  const uint32_t kRequestLength = 2;
+  const uint64_t requestBuffer[kRequestLength] = {(uint64_t)200, expectedLastValue};
+  ViewNum curView = 0;
+  bftEngine::impl::SeqNum AssignedSeqNum = 1;
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: nit - `assignedSeqNum`

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 6]
```diff
@@ -51,6 +135,99 @@ void setUpConfiguration_4() {
...
+
+  auto VCS = ViewChangeSafetyLogic(
+      N, F, C, replicaConfig[0].thresholdVerifierForSlowPathCommit, PrePrepareMsg::digestOfNullPrePrepareMsg());
+
+  SeqNum min{}, max{};
+  VCS.computeRestrictions(viewChangeMsgs, 0, min, max, restrictions);
+
+  for (int i = 0; i < kWorkWindowSize; i++) {
+    if (i == AssignedSeqNum - 1) {
+      Assert(restrictions[i].isNull == false);
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: doesn't it work with `Assert(!restrictions[i].isNull)` ?

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 7]
```diff
@@ -51,6 +135,99 @@ void setUpConfiguration_4() {
...
+      Assert(pp->digestOfRequests().toString() == restrictions[i].digest.toString())
+    } else {
+      Assert(restrictions[i].isNull == true);
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: same here `Assert(restrictions[i].isNull)` ?

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 8]
```diff
@@ -99,6 +276,30 @@ TEST(testViewchangeSafetyLogic_test, different_new_views_in_VC_msgs) {
...
+    } else {
+      viewChangeMsgs[i] = new ViewChangeMsg(i, 1, 0);
+    }
+  }
+
+  auto VCS = ViewChangeSafetyLogic(
+      N, F, C, replicaConfig[0].thresholdVerifierForSlowPathCommit, PrePrepareMsg::digestOfNullPrePrepareMsg());
+
+  auto seqNum = VCS.calcLBStableForView(viewChangeMsgs);
+  Assert(seqNum == 0);
```
Comment (teoparvanov) [inline bftengine/tests/testViewChange/testViewChange.cpp]: `seqNum` 0 doesn't make sense IMO (to me it looks more like a corner case)

File: bftengine/tests/testViewChange/testViewChange.cpp [hunk 9]
```diff
@@ -51,6 +135,99 @@ void setUpConfiguration_4() {
...
+  auto operation_status = replica.combine_operation_future.get();
+  assert(operation_status == DummyReplica::PrepareCombinedSigOperationStatus::OPERATION_SUCCEEDED);
+  auto pfMsg = seqNumInfo_.getValidPrepareFullMsg();
+
+  ViewChangeMsg** viewChangeMsgs = new ViewChangeMsg*[N];
+  for (int i = 0; i < N; i++) {
+    if (i == 1) {
+      viewChangeMsgs[i] = nullptr;
+    } else {
+      viewChangeMsgs[i] = new ViewChangeMsg(i, 1, 0);
```
[the remark this answers is not in the document]
Comment (HristoStaykov) [reply bftengine/tests/testViewChange/testViewChange.cpp]: Yes, just have to be careful to set assignedSeqNum > lastStableSeq, otherwise assignedSeqNum will be ignored.

**Comment by teoparvanov** [comment]
It looks great, @HristoStaykov :)
I think you can remove the "WIP" from the PR title now.
